### ISSUE TITLE
Positional Sound

### ISSUE THREAD

**Issue by Jaytheway** [TeamPorcupine/ProjectPorcupine#1576]
### Expected behavior

Sound should play from the position of its emitters.
Unity has this capabilities. Not sure to what extent they're applicable in our situation.

- **Min Distance**	Within the MinDistance, the sound will stay at loudest possible. Outside MinDistance it will begin to attenuate. Increase the MinDistance of a sound to make it ‘louder’ in a 3d world, and decrease it to make it ‘quieter’ in a 3d world.
- **Max Distance**	The distance where the sound stops attenuating at. Beyond this point it will stay at the volume it would be at MaxDistance units from the listener and will not attenuate any more.
- **Rolloff Mode**	How fast the sound fades. The higher the value, the closer the Listener has to be before hearing the sound. (This is determined by a Graph).
- **Logarithmic Rolloff**	The sound is loud when you are close to the audio source, but when you get away from the object it decreases significantly fast.
- **Linear Rolloff**	The further away from the audio source you go, the less you can hear it.
- **Custom Rolloff**	The sound from the audio source behaves accordingly to how you set the graph of roll offs.

As I understand, Custom Rolloff is not an option.

Unity Manual page for the reference: https://docs.unity3d.com/Manual/class-AudioSource.html
Unity Scripting API: https://docs.unity3d.com/ScriptReference/AudioSource.html

### Actual behavior

For now sounds play exactly in the center (in a 2D space (where the camera is)) ignoring position of the emitter relatively to the camera.

**Examples**: 
- If aliens attack from the north, you would here sound directly in your face.
- If the fire started you would here the sound in your face, without any directional indication.
- All furniture placements - in your face.

All these resolves not only in an aesthetic nonsense, but also in a muddy mix. (Especially without any means to mix audio) 

### Steps to reproduce the behavior

- Step 1: Get positions of sound emitter and camera (perhaps, the tile, which is directly below the camera)
- Step 2: Calculate the distance between sound emitter and camera 
- Step 3: ...

Also in Uniti Scritping API there is: "You can play a clip at a static position in 3D space using PlayClipAtPoint." It might be a way to do it(?)

### Related issues

This is one of the basic features that our sound system need listed here: #1535

I am by no means a programmer, so: 
-If you know how to do it
-If you are willing to put time in something that will determine how the game feels
-If you feel strong about it, as I am listening to the "[Nerevar Rising](https://www.youtube.com/watch?v=QuBB0JchGZc)" from the Morrowind's soundtrack!

### Put your name on the pedestal of glory!

# Pedestal of Glory
(names of all contributors to the Sound Engine will be listed here)

**Comment by koosemose** [comment]
I'm working on replacing the core sound system (due to problems with streamingAssets and the unity sound system), it may make some of these things easier, I'm not sure yet, currently I just have it mimicing the current functionality. I can give more details once I get further into the work.

**Comment by bjubes** [comment]
If you want custom rolloff, you can use unity's [animation curve](https://docs.unity3d.com/ScriptReference/EditorGUILayout.CurveField.html) to set up a custom graph of volume v distance (maybe save it in a Scriptable Object/prefab or something?). Not moddable, but something like a sound system roll off doesn't really have to be IMO

**Comment by koosemose** [comment]
By default we have access to INVERSEROLLOFF, LINEARROLLOFF, LINEARSQUAREROLLOFF, CUSTOMROLLOFF, which, essentially can be applied individually per sound, for the sound system as a whole, or a particular playing of a sound. Default is Inverse Rolloff. Custom Rolloff takes a list of vectors (technically 3d vectors but it ignores the z). I don't really understand why we would need different sounds to roll off differently, shouldn't all sounds in the world roll off the same? I'm just trying to figure out the use cases so I don't waste time programming some aspect we don't really have a need for, as the normal syntax is a bit obtuse, so I'm working on more clear methods to allow us to do what we need to do without having to use a bunch of obtuse code, except for in the in-between methods I'm working on.

**Comment by Jaytheway** [comment]
Different types of sound degrade differently when you getting farther away from them.  
Some ambient sounds feel more natural when they rolloff linearly, wheres others, like furniture placements, would feel unnatural if volume attenuation would be the same at all distances.
When Stars colliding on the background, you hear them equally loud wherever you are in Space. But at some ginormous distance you'd not hear them.

Human hearing is non linear by nature, it's more like a logarithmic curve. That's why usually logarithmic and S curved rolloffs feel more natural.

Not to mention the lengths of sound waves for different frequencies... That level of realism we don't need. Although it wouldn't be too difficult to implement (High Freq Cutoff to Distance).

**Comment by koosemose** [comment]
For now I will probably leave roll off as the default (which seems to be logarithmic) due to the amount of supporting code that would be required.

I'm trying to figure out a reasonable way to make min/max distance accessible, because I'm currently using min to basically increase the area at which sound is not attenuating, because with the omniscient viewpoint it feels very odd for things that are a little bit to the left the center of the viewpoint to be as attenuated as they become.

**Comment by Jaytheway** [comment]
Yeah, I figured that camera's view point would screw some things up.
Is there a possibility to calculate distance not from the tile below the camera, but from the camera itself to the object in a 3D space (x, y, z)? 
So when the camera zooms out, sound attenuates accordingly. And attenuation behave more naturally.

**Comment by koosemose** [comment]
Calculating from the camera's position will give worse results, in several ways. The camera's not where you think it is, it is ALWAYS 10 units back from the center tile, you don't really zoom in and out, the field of view (orthographic size) just gets larger or smaller. Now, I could fake it by setting the listener to a point some amount of space back from the center tile and vary that by a function of the orthographic size, but that will bring it's own problems, if, for example, you are viewing the bottom most layer, and I've set the listener to an amount back higher than 1 unit you're going to hear sounds from a level (or more depending on how much higher than 1) above your current level as close. 

So the answer is, yes I could, but it would lead to worse results.

**Comment by Geoffrotism** [comment]
Would it make sense to have all sounds play at the position of the camera, but then use panning and geommetry to determine how loud and how much in each ear we want to have the audio play? we would of course use the distance from the tile making the sound to the tile that the camera is technically looking at.

**Comment by koosemose** [comment]
o.O That would just be manually doing positional audio... I'm not sure I see how that is in any way a benefit.

**Comment by bjubes** [comment]
IMO audio should not be heavily effected by where the camera is (3d audio) since it is meant for a 3d fps style game, not a 2d top down one. it gives the impression that the game is happening around the player which doesn't align with the player seeing the entire game as a flat surface in front of them.
